src/support/modify.c

```diff
@@ -179,165 +221,129 @@ __modify_apply_one(
-	    ((char *)value->data)[value->size - 1] == '\0');
+		/* Step over the current modification. */
+		destoff += current.data.size;
 
-	*successp = true;
+		prev = current;
+		first = false;
+	} WT_MODIFY_FOREACH_END;
 
-err:	__wt_scr_free(session, &tmp);
-	return (ret);
+	/* Step over the final unmodified block. */
+	destoff += value->size - (current.offset + current.size);
 
+	*dataszp = datasz;
+	*destszp = destoff;
+	return (true);
 }
 
 /*
- * __modify_apply --
- *	Apply a single set of WT_MODIFY changes to a buffer.
+ * __modify_apply_no_overlap --
+ *	Apply a single set of WT_MODIFY changes to a buffer, where the changes
+ *	are in sorted order and none of the changes overlap.
  */
 static int
-__modify_apply(WT_SESSION_IMPL *session,
-    WT_CURSOR *cursor, WT_MODIFY *entries, int nentries)
+__modify_apply_no_overlap(WT_SESSION_IMPL *session,
+    WT_ITEM *value, const size_t *p, int nentries, size_t datasz, size_t destsz)
 {
-	WT_ITEM *value;
-	int i;
-	bool success;
-
-	value = &cursor->value;
+	WT_MODIFY current;
+	size_t sz;
+	const uint8_t *from;
+	uint8_t *to;
+	int delta;
 
 	/*
 	 * Grow the buffer first. This function is often called using a cursor
 	 * buffer referencing on-page memory and it's easy to overwrite a page.
 	 * A side-effect of growing the buffer is to ensure the buffer's value
 	 * is in buffer-local memory.
 	 */
-	WT_RET(__wt_buf_grow(session, value, value->size));
-
-	/*
-	 * Try fast path #1.
-	 * Do simple replacement in the existing buffer if the modifications are
-	 * overwriting fixed-sized fields.
-	 *
-	 * If either the data size or replacement count extend past the end of
-	 * the current value, we have to deal with padding bytes. Don't try to
-	 * fast-path padding bytes; it's not common and adds branches to the
-	 * loop applying the changes.
-	 */
-	for (i = 0; i < nentries; ++i) {
-		if (entries[i].data.size != entries[i].size ||
-		    value->size < entries[i].offset + entries[i].size)
-			break;
-
-		memmove((uint8_t *)value->data + entries[i].offset,
-		   entries[i].data.data, entries[i].data.size);
-	}
-
-	/* Check for completion, else, skip any entries consumed. */
-	if (i == nentries)
-		return (0);
-	nentries -= i;
-	entries += i;
+	WT_RET(__wt_buf_grow(session, value, WT_MAX(destsz, value->size)));
+
+	delta = 0;
+	WT_MODIFY_FOREACH_BEGIN(current, p, nentries) {
+		delta += (int)current.data.size - (int)current.size;
+	} WT_MODIFY_FOREACH_END;
+
+	from = (const uint8_t *)value->data + value->size;
+	to = (uint8_t *)value->mem + destsz;
+	WT_MODIFY_FOREACH_REVERSE(current, p, nentries, datasz) {
+		/* Move the current unmodified block into place if necessary. */
+		sz = WT_PTRDIFF(to, value->mem) -
+		    (current.offset + current.data.size);
+		from -= sz;
+		to -= sz;
+		WT_ASSERT(session, from >= (const uint8_t *)value->data &&
+		    to >= (uint8_t *)value->mem);
+		WT_ASSERT(session,
+		    from + sz <= (const uint8_t *)value->data + value->size);
```

Comment:
I think the previous version of this code should remain. If we're overwriting fixed-size fields, we don't need to do anything other than a copy, and the check is worth doing because we can simply consume any number of changes until we hit one that isn't fixed size.

Reply:
Done.  It's a little messy to keep track of a partially applied modify, but in the common case where everything applies in place there is now a single pass.